libs/ui/src/hooks/smartcard_auth/auth_helpers.test.ts

```diff
@@ -191,6 +205,396 @@ describe('Card interface', () => {
+    await card.unprogramUser();
+    jest.advanceTimersByTime(CARD_POLLING_INTERVAL);
+    await waitForNextUpdate();
+    expect(result.current.card).toBeDefined();
+    card = result.current.card!;
+    expect(card.programmedUser).not.toBeDefined();
+    expect(card.hasStoredData).toEqual(false);
+    expect(logger.log).toHaveBeenNthCalledWith(
+      12,
+      LogEventId.SmartcardUnprogramInit,
+      'superadmin',
+      {
+        message: 'Unprogramming pollworker smartcard...',
+        programmedUserRole: 'pollworker',
+      }
+    );
+    expect(logger.log).toHaveBeenNthCalledWith(
+      13,
+      LogEventId.SmartcardUnprogramComplete,
+      'superadmin',
+      {
+        disposition: 'success',
+        message: 'Successfully unprogrammed pollworker smartcard.',
+        previousProgrammedUserRole: 'pollworker',
+      }
+    );
+
+    // Unprogram the card again to verify that unprogramming an already unprogrammed card doesn't
+    // cause problems
+    await card.unprogramUser();
+    jest.advanceTimersByTime(CARD_POLLING_INTERVAL);
+    await waitForNextUpdate();
+    expect(result.current.card).toBeDefined();
+    card = result.current.card!;
+    expect(card.programmedUser).not.toBeDefined();
+    expect(card.hasStoredData).toEqual(false);
+    expect(logger.log).toHaveBeenNthCalledWith(
+      14,
+      LogEventId.SmartcardUnprogramInit,
+      'superadmin',
+      {
+        message: 'Unprogramming unprogrammed smartcard...',
+        programmedUserRole: 'unprogrammed',
+      }
+    );
+    expect(logger.log).toHaveBeenNthCalledWith(
+      15,
+      LogEventId.SmartcardUnprogramComplete,
+      'superadmin',
+      {
+        disposition: 'success',
+        message: 'Successfully unprogrammed unprogrammed smartcard.',
+        previousProgrammedUserRole: 'unprogrammed',
+      }
+    );
+  });
+
+  it('raises an error on concurrent programming requests', async () => {
+    const cardApi = new MemoryCard();
+    const logger = fakeLogger();
+    const { result, waitForNextUpdate } = renderHook(() =>
+      useDippedSmartcardAuth({ cardApi, logger })
+    );
+    authAsSuperAdmin(cardApi);
+    await waitForNextUpdate();
+    assert(isSuperadminAuth(result.current));
+
+    // Insert an unprogrammed card
+    cardApi.insertCard();
+    jest.advanceTimersByTime(CARD_POLLING_INTERVAL);
+    await waitForNextUpdate();
+    expect(result.current.card).toBeDefined();
+    const card = result.current.card!;
+
+    // Make concurrent programming requests
+    const [write1, write2] = await Promise.all([
+      card.programUser({ role: 'superadmin' }),
+      card.programUser({ role: 'pollworker', electionHash }),
+    ]);
+    expect(
+      (write1.isErr() && !write2.isErr()) || (!write1.isErr() && write2.isErr())
+    ).toEqual(true);
+    if (write1.isErr()) {
+      expect(write1.err()?.message).toEqual('Card write in progress');
+    } else {
+      expect(write2.err()?.message).toEqual('Card write in progress');
+    }
```

Comment:
Did you actually see these get executed in different orders? Or does the same one always win? Just curious

Reply:
I didn't! The first in the `Promise.all` array always seems to get the lock first

I know making that assumption would simplify this test (I saw you made this assumption with the other concurrency tests in this file). But I ultimately decided to play it safe since it still seems theoretically possible for the second to get the lock first